src/core/connection_pool.c

```diff
@@ -0,0 +1,508 @@
+
+    for (uint32_t i = 0; i < Config->NumberOfConnections; i++) {
+        const uint32_t MaxCreationRetries = RssProcessorCount * MAX_CONNECTION_POOL_RETRY_MULTIPLIER;
+        uint32_t RetryCount = 0;
+        uint32_t RssProcIndex;
+        for (; RetryCount < MaxCreationRetries; RetryCount++) {
+
+            uint32_t NewPort = QuicAddrGetPort(&LocalAddress) + 1;
+            if (NewPort > QUIC_ADDR_EPHEMERAL_PORT_MAX) {
+                NewPort = QUIC_ADDR_EPHEMERAL_PORT_MIN;
+            }
+            QuicAddrSetPort(&LocalAddress, (uint16_t)NewPort);
+
+            uint32_t RssHash = 0, Offset;
+            //
+            // Calculate the Toeplitz Hash as if receiving packets from the
+            // ResolvedRemoteAddress to find the RSS processor.
+            //
+            CxPlatToeplitzHashComputeRss(
+                &ToeplitzHash,
+                &ResolvedRemoteAddress,
+                &LocalAddress,
+                &RssHash,
+                &Offset);
+            uint32_t Mask = (RssConfig->RssIndirectionTableLength / sizeof(uint32_t)) - 1;
+            RssProcIndex = RssProcessorCount;
+            CXPLAT_DBG_ASSERT((RssHash & Mask) < RssConfig->RssIndirectionTableLength / sizeof(uint32_t));
+            uint32_t RssProc = RssConfig->RssIndirectionTable[RssHash & Mask];
+            for (uint32_t j = 0; j < RssProcessorCount; j++) {
+                if (RssProcessors[j] == RssProc) {
+                    RssProcIndex = j;
+                    break;
+                }
+            }
+            CXPLAT_DBG_ASSERT(RssProcIndex < RssProcessorCount);
+            if (ConnectionCounts[RssProcIndex] >= ConnectionsPerProc) {
+                //
+                // This processor already has enough connections on it, so try another port number.
+                //
+                continue;
+            }
+
+            uint16_t PartitionIndex =
+                QuicLibraryGetPartitionFromProcessorIndex(RssProcessors[RssProcIndex]);
+
+            Status =
+                QuicConnAlloc(
+                    (QUIC_REGISTRATION*)Config->Registration,
+                    NULL,
+                    NULL,
+                    &PartitionIndex,
+                    &Connections[i]);
+            if (QUIC_FAILED(Status)) {
+                QuicTraceLogError(
+                    ConnPoolOpenConnection,
+                    "[conp] Failed to open connection[%u], 0x%x",
+                    i,
+                    Status);
+                goto Error;
+            }
+            CreatedConnections++;
+            Connections[i]->ClientCallbackHandler = Config->Handler;
+            if (Config->Context != NULL) {
+                Connections[i]->ClientContext = Config->Context[i];
+            }
+
+            //
+            // Set parameters on the connections before starting them.
+            //
+            if (Config->CibirIds) {
+                uint8_t True = TRUE;
+                Status =
+                    QuicConnParamSet(
```

Comment:
Part of me thinks it's overkill calling QuicConnParamSet for each of these, instead of just setting the parameters directly.

Reply:
`QuicConnParamSet` does some parameter checks, which we need anyway in case the caller passed bad info, before copying the info (particularly for CIBIR).  That doesn't seem like overkill to me.
Sure I could copy the local and remote addresses directly if it seems like such a burden, but I think it keeps the code looking clearer than manually copying.